### PR TITLE
Add Mixtral Decode for 32K seqlen to main

### PR DESCRIPTION
**Pull request by caixunshiren**
### Description
With the support of flash decode v2, we can enable decode to up to 32k with batch 16, and up to 16k with batch 32.

### Known Issue
It has been observed that e2e perf for small seqlen (<1k) dropped from 14 tok/s to 13 tok/s with this new change. Large seqlen has relative stable perf, with around 11 tok/s for 8k, 10.5 tok/s for 16k and 9.5 tok/s for 32k respectively.

It it suspected that the degraded e2e perf is due to the complicated update runtime args function in flash decode. A separate P1 issue is opened to track it down #10496.

As per discussion in the all hand meeting, we would want functionality over perf. Therefore this PR is opened merge the 32k seqlen functionality into main.

### CODE REVIEW

**Review by yieldthought — APPROVED** [review]
Love all that code replaced with the sdpa op call ^^

**Comment by yieldthought** [comment]
Fired off some tests, please check these pass before merging:
 [ ] [Unit tests](https://github.com/tenstorrent/tt-metal/actions/runs/10020737608)
 [ ] [Perf tests](https://github.com/tenstorrent/tt-metal/actions/runs/10020744578)
 [ ] [Post commit](https://github.com/tenstorrent/tt-metal/actions/runs/10020746727)
 [ ] [Perplexity](https://github.com/tenstorrent/tt-metal/actions/runs/10020748691)
 [ ] [Demo](https://github.com/tenstorrent/tt-metal/actions/runs/10020750177)

File: models/demos/t3000/mixtral8x7b/tt/mixtral_attention.py
Comment (mtairum) [inline models/demos/t3000/mixtral8x7b/tt/mixtral_attention.py:164]: for forward_decode we just pass a single rot_mat. For correctness, can you re-add the correct description? `rot_mat: rotation matrix for each device`.

Can you also add the input description to the `forward_prefill` function for completeness? 🙏

**Review by mtairum — CHANGES_REQUESTED** [review]

File: models/demos/t3000/mixtral8x7b/tt/mixtral_common.py
Comment (mtairum) [inline models/demos/t3000/mixtral8x7b/tt/mixtral_common.py:171]: Add a comment to this function to explain where/what for we use these chunk sizes

**Review by mtairum — CHANGES_REQUESTED** [review]

File: models/demos/t3000/mixtral8x7b/tt/mixtral_common.py
Comment (caixunshiren) [reply models/demos/t3000/mixtral8x7b/tt/mixtral_common.py:171]: Removed this function because it is no longer needed (it was used in flash attention v1)

**Comment by caixunshiren** [comment]
Looks like the demo and perf tests failed because I didn't apply my changes in files other than demo.py and test_mixtral_model.py. Fixing this issue rn ...

**Comment by caixunshiren** [comment]
New tests fired after fix:

* [ ]  [Unit tests](https://github.com/tenstorrent/tt-metal/actions/runs/10046564844)
* [ ]  [Perf tests](https://github.com/tenstorrent/tt-metal/actions/runs/10046568348)
* [ ]  [Post commit](https://github.com/tenstorrent/tt-metal/actions/runs/10046579211)
* [ ]  [Perplexity](https://github.com/tenstorrent/tt-metal/actions/runs/10046582371)
* [ ]  [Demo](https://github.com/tenstorrent/tt-metal/actions/runs/10046588384)

**Comment by caixunshiren** [comment]
> New tests fired after fix:
> 
> * [ ]   [Unit tests](https://github.com/tenstorrent/tt-metal/actions/runs/10046564844)
> * [ ]   [Perf tests](https://github.com/tenstorrent/tt-metal/actions/runs/10046568348)
> * [ ]   [Post commit](https://github.com/tenstorrent/tt-metal/actions/runs/10046579211)
> * [ ]   [Perplexity](https://github.com/tenstorrent/tt-metal/actions/runs/10046582371)
> * [ ]   [Demo](https://github.com/tenstorrent/tt-metal/actions/runs/10046588384)

@mtairum the output of demo does not match the exact tokens saved in CI, but the output still makes sense. What should we do about it?

**Comment by yieldthought** [comment]
Agreed that the demo output exact match is tripping us up here, the output matches for a really large number of tokens and then diverges in a way that looks absolutely fine for user 3 at least.

**Comment by mtairum** [comment]
> the output matches for a really large number of tokens and then diverges in a way that looks absolutely fine for user 3 at least.

Keep in mind that the majority of those tokens is prefilled from the prompt - only the last 50 tokens are actually generated, and for user 3 it looks ok. 

@caixunshiren I would comment out the assert and print out all the output mismatches and fix the expected outputs manually to account for this new version, assuming that all other mismatches still look good enough.

**Comment by caixunshiren** [comment]
> Perf test fail is weird:
> 
> 
> 
> - kv_cache_len=32: 0.0738 = 13.6 t/s/u
> 
> - kv_cache_len=128: 0.1025 = 9.8 t/s/u
> 
> - kv_cache_len=1024: 0.0783 = 12.8 t/s/u
> 
> - kv_cache_len=2048: 0.0793 = 12.6 t/s/u
> 
> 
> 
> I don't have any theories as to why a shorter kv-cache length would be 3-4 t/s/u slower - that's over 30%! Is this reproducible locally? Is there that much noise in our perf measurements now?
> 
> 
> 
> Fired off a second run to compare and see if we get the same figures again: https://github.com/tenstorrent/tt-metal/actions/runs/10056248458

Yes I can confirm the variation is big on local. Seems to be not a flash decode issue because I also see the exact same problem with current main branch, where the perf can drop down to 10 tok/s/u in some cases.

**Comment by yieldthought** [comment]
Rerun perf test passes. 🤔 we should really try to decrease the variance on this, but as you point out that should not block this PR.

**Comment by caixunshiren** [comment]
> Rerun perf test passes. 🤔 we should really try to decrease the variance on this, but as you point out that should not block this PR.

Should we relax the constraint? Also, I actually reran the perf test after you started it... the first time timedout at 300 seconds for some reason... maybe increase the timeout as well?

**Comment by caixunshiren** [comment]
In the latest commit, I changed the demo ground truth output to match the new version. I also increased the timeout of perf test from 300 seconds to 400 seconds, as I observed timeout 2 times during previous run and I had to rerun the pipeline. I believe the time for the test to run is right at the boundary.

* [ ] [demo test ](https://github.com/tenstorrent/tt-metal/actions/runs/10065343606)
* [ ] [perf test](https://github.com/tenstorrent/tt-metal/actions/runs/10065352132)

**Comment by mtairum** [comment]
Increasing timeout is ok.

Variance we have observed that in the past both locally and CI. We have to account for worst case scenario and use a worse estimate, otherwise we risk CI pipelines to fail frequently.

The latest runs look good.
The normal `demo.py` asserted on first expected output. You'll have to updated the expected one for it.

**Review by mtairum — APPROVED** [review]

**Comment by caixunshiren** [comment]
Launching last set of test before merging:

- [ ] [Unit tests](https://github.com/tenstorrent/tt-metal/actions/runs/10113228168)
- [ ] [Perf tests](https://github.com/tenstorrent/tt-metal/actions/runs/10113249669)
- [ ] [Post commit](https://github.com/tenstorrent/tt-metal/actions/runs/10113261702)
- [ ] [Demo](https://github.com/tenstorrent/tt-metal/actions/runs/10113238831)

**Comment by caixunshiren** [comment]
All tests passed. Merging to Main

**Pull request merged**
